kolibri/plugins/coach/assets/src/composables/useResourceSelection.js

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * @typedef {import('../../../../../../composables/useFetch').FetchObject} FetchObject
+ */
+
+/**
+ * Composable for managing the selection of resources within a topic tree.
+ * This utility handles selection rules, manages fetch states for channels, bookmarks,
+ * and topic trees, and offers methods to add, remove, or override selected resources.
+ *
+ * @typedef {Object} UseResourceSelectionResponse
+ * @property {Object} topic Topic tree object, contains the information of the topic,
+ *   its ascendants and children.
+ *   Defined only if the `topicId` query in the route is set.
+ * @property {boolean} loading Indicates whether the main topic tree, channels, and bookmarks
+ *   data are currently loading. This does not account for loading more data. For such cases,
+ *   use the fetch objects of each entity.
+ * @property {FetchObject} channelsFetch Channels fetch object to manage the process of
+ *   fetching channels. We currently don't support fetching more channels.
+ * @property {FetchObject} bookmarksFetch Bookmarks fetch object to manage the process of
+ *   fetching bookmarks. Fetching more bookmarks is supported.
+ * @property {FetchObject} treeFetch Topic tree fetch object to manage the process of
+ *   fetching topic trees and their resources. Fetching more resources is supported.
+ * @property {Array<(node: Object) => boolean>} selectionRules An array of functions that determine
+ *   whether a node can be selected.
+ * @property {Array<Object>} selectedResources An array of currently selected resources.
+ * @property {(resources: Array<Object>) => void} selectResources Adds the specified resources
+ *   to the `selectedResources` array.
+ * @property {(resources: Array<Object>) => void} deselectResources Removes the specified resources
+ *   from the `selectedResources` array.
+ * @property {(resources: Array<Object>) => void} setSelectedResources Replaces the current
+ *   `selectedResources` array with the provided resources array.
+ *
+ * @returns {UseResourceSelectionResponse}
+ */
+export default function useResourceSelection() {
+  const store = getCurrentInstance().proxy.$store;
+  const route = computed(() => store.state.route);
+  const topicId = computed(() => route.value.query.topicId);
+
+  const selectionRules = ref([]);
+  const selectedResources = ref([]);
+
+  const bookmarksFetch = useFetch({
+    fetchMethod: () =>
+      ContentNodeResource.fetchBookmarks({
+        params: { limit: 25, available: true },
+      }),
+    fetchMoreMethod: more =>
+      ContentNodeResource.fetchBookmarks({
+        params: more,
+      }),
+    dataKey: 'results',
+    moreKey: 'more',
+    additionalDataKeys: {
+      count: 'count',
+    },
+  });
+
+  const channelsFetch = useFetch({
+    fetchMethod: () =>
+      ChannelResource.fetchCollection({
+        getParams: {
+          available: true,
+        },
+      }),
+  });
+
+  const treeFetch = useFetch({
+    fetchMethod: () =>
```

Comment:
I think we could simplify the abstractions in useFetch by dealing with the special case for tree fetching in here instead.

We can create a separate `topic` ref that we assign the whole result of `ContentNodeResource.fetchTree` to, and then return:
```
ContentNodeResource.fetchTree({ id: topicId.value, params: { include_coach_content: true } }).then(t => {
   topic.value = t;
   return topic.children;
});
```
This allows us to avoid any special casing, and the returned data is still a `more` parameter, so we don't need to customize the `moreKey` either, and that can be static.

Reply:
It now looks so much better :)

```javascript
  const fetchTree = async (params = {}) => {
    topic.value = await ContentNodeResource.fetchTree(params);
    return topic.value.children;
  };

  const treeFetch = useFetch({
    fetchMethod: () => fetchTree({ id: topicId.value, params: { include_coach_content: true } }),
    fetchMoreMethod: more => fetchTree(more),
    dataKey: 'results',
    moreKey: 'more',
  });
```

Reply:
Do we need `dataKey` and `moreKey` to be configurable now? I have deliberately made all the API endpoints use consistent return values for these, so I think it shouldn't need to be!

Reply:
Are them? Wow, Just for the answer of the channels endpoint I thought they were not. Because there the data is the whole response object. But I guess in that case different responses are because that response is not paginated right? So we can assume that if we dont send a "fetchMoreMethod", then the `data` is just the whole response object, like the channels response. And if we do send a "fetchMoreMethod", then the data will always come in "response.results", the more object in "response.more", and the count in "response.count", right?